src/include/transaction/deferred_action_manager.h

```diff
@@ -91,20 +104,51 @@ class DeferredActionManager {
     }
   }
 
+  /**
+   * // TODO(John, Ling): Eventually we should remove the special casing of indexes here. See processIndexes()
+   * Register an index to be periodically garbage collected
+   * @param index pointer to the index to register
+   */
+  void RegisterIndexForGC(common::ManagedPointer<storage::index::Index> index);
+
+  /**
+   * // TODO(John, Ling): Eventually we should remove the special casing of indexes here. See processIndexes()
+   * Unregister an index to be periodically garbage collected
+   * @param index pointer to the index to unregister
+   */
+  void UnregisterIndexForGC(common::ManagedPointer<storage::index::Index> index);
+
+  /**
+   * @return Pointer to the visited tuple slot set
+   */
+  std::unordered_set<storage::TupleSlot> *GetVisitedSlotsLocation() { return &visited_slots_; }
+
  private:
   const common::ManagedPointer<TimestampManager> timestamp_manager_;
   // TODO(Tianyu): We might want to change this data structure to be more specialized than std::queue
-  std::queue<std::pair<timestamp_t, DeferredAction>> new_deferred_actions_, back_log_;
-  common::SpinLatch deferred_actions_latch_;
+  tbb::concurrent_queue<std::pair<timestamp_t, DeferredAction>> new_deferred_actions_;
+  std::queue<std::pair<timestamp_t, DeferredAction>> back_log_;
+  // It is sufficient to truncate each version chain once in a GC invocation because we only read the maximal safe
+  // timestamp once, and the version chain is sorted by timestamp. Here we keep a set of slots to truncate to avoid
+  // wasteful traversals of the version chain.
+  std::unordered_set<storage::TupleSlot> visited_slots_;
```

Comment:
This seems like something that should belong in `garbage_collector.h`. I feel kind of iffy with having special cased gc logic in the deferred action manager. Ideally the deferred action manager is dumb and only processes the actions registered, so it wouldn't know if those actions are truncating an already visited_slot.

Reply:
Eventually the gc will be removed though. Where do you suggest to place visited_slots_?

Reply:
I'm not sure if the gc should be removed. I think the garbage collector thread should be removed. But the `garbage_collector.cpp` should still have gc functions that execute gc specific logic. These functions would be called in deferred events but the logic of these functions shouldn't be a member or visible to deferred_action_manager. In the past that used to be `truncate_version_chain` and such but in this pr that's been moved to be a member of `DataTable` which is fine and make sense to me. So `garbage_collector.cpp` will probably end up not having much but I feel like gc specific logic should remain there.